content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+integrate it or write a custom step that will invoke it.
+.  *Assume things will go wrong:* don’t rely on workspaces being clean
+of the remnants from previous executions, clean explicitly where needed.
+Make use of timeouts and retry steps (that’s what they’re there for).
+    ..  Within a git repository, ‘git clean -fdx’ is a good way to
+accomplish this and reduces the amount of SCM cloning
+.  *DO use parameterized pipelines and variables to make your pipeline
+scripts more reusable.* Passing in parameters is especially helpful for
+handling different environments and should be preferred to applying
+conditional lookup logic.
+.  *Try to limit business logic embedded in pipelines.* To some extent
+this is inevitable, but try to focus on tasks to complete instead,
+because this yields more maintainable, reusable, and often more
+performant pipeline code.
+    ..  One code smell that points to a problem is many hard-coded
+constants. Consider taking advantage of the options above to refactor
+code for better composability.
+    ..  For complex cases, consider using Jenkins integration options
+(plugins, Jenkins API calls, invoking input steps externally) to offload
+implementation of more complex business rules to an external system if
+they fit more naturally there.
+
+Please, think of these as guidelines, not strict rules – Jenkins
+pipeline provides a great deal of power and flexibility, and it's there
+to be used.
+
+Breaking enough of these rules at scale can cause masters to fail by
+placing an unsustainable load on them.
+
+For additional guidance, we also recommend the Jenkins World talk on how
+to engineer pipelines for speed and performance:
+https://www.cloudbees.com/need-speed-building-pipelines-be-faster
+
+[[appendix-serializable-vs.-non-serializable-types]]
+Appendix: Serializable vs. Non-Serializable Types:
+--------------------------------------------------
+
+To assist with pipeline development, here are common serializable and
+non-serializable types, to assist with deciding if your logic can be CPS
+or should be in a @NonCPS function to avoid issues.
+
+*Common Serializable Types (safe everywhere):*
+
+.  All primitive types and their object wrappers: byte, boolean, int,
+double, short, char
+.  Strings
+.  enums
+.  Arrays of serializable types
+.  ArrayLists and normal Groovy Lists
+.  Sets: HashSet
+.  Maps: normal Groovy Map, HashMap, TreeMap
+.  Exceptions
+.  URLs
+. Dates
+. Regex Patterns (compiled patterns)
+
+*Common non-Serializable Types (only safe in @NonCPS functions):*
+
+. Iterators: this is a common problem. You need to use C-style loop, i.e.
+`for(int i=0; i<max; i++)\{`
+. Regex Matchers (you can use the
+built-in functions in String, etc, just not the Matcher itself)
+. *Important:* JsonObject, JsonSlurper, etc in Groovy 2+ (used in some 2.x+
+versions of Jenkins).
+..  This is due to an internal implementation change
+-- earlier versions may serialize.
```

Comment:
Really? Which? There was no implementation change I know of affecting these types. They should just not be used generally. Too fragile, too hard to secure.